packages/aws-client/src/S3.ts

```diff
@@ -342,34 +353,185 @@ export const s3PutObjectTagging = improveStackTrace(
+ *    `DEFAULT_RETRY_OPTIONS`
+ */
+const retryable = <P, D>(s3Fn: S3RequestFunction<P, D>) =>
+  (params: P, options?: pRetry.Options) =>
+    pRetry(() => s3Fn(params), { ...DEFAULT_RETRY_OPTIONS, ...options });
+
+/**
+ * Makes an S3 function retryable _only_ for the specified list of error codes.
+ * Returns a function that wraps an S3 function such that the wrapper behaves
+ * identically to the wrapped S3 function, except that all errors with status
+ * codes matching one of the specified status codes are directly rethrown, while
+ * all other errors are wrapped in a `pRetry.AbortError` before being thrown.
+ *
+ * @example
+ * const abortableGetObject = retryStatusCodes([304, 404, 412])(getObject);
+ * // Throws pRetry.AbortError if getObject throws an error with a `statusCode`
+ * // that is NOT in the list of status codes above.
+ * const result = await abortableGetObject({ Bucket, Key });
+ *
+ * @function
+ * @param {number[]} statusCodes - status codes of errors that should allow for
+ *    retrying a request; all other errors are wrapped in a `pRetry.AbortError`
+ * @returns {function} a function that wraps an S3 function such that all errors
+ *    thrown by the wrapped function that have status codes other than the
+ *    specified status codes are wrapped in `pRetry.AbortError`s (i.e., only
+ *    the specified errors result in retrying the failed request when the
+ *    wrapper function is used in conjunction with `pRetry()`)
+ */
+const retryStatusCodes = (statusCodes: number[]) =>
+  <P, D>(s3Fn: S3RequestFunction<P, D>) =>
+    (params: P) =>
+      s3Fn(params).catch((error) => {
+        if (statusCodes.includes(error.statusCode)) throw error;
+        throw new pRetry.AbortError(error);
+      });
+
+/**
+ * Returns a function that accepts a resolved promise from an S3 function.  The
+ * returned function checks the resolved value against the pre-conditions
+ * specified in `params`, and either returns the resolved value, if the
+ * pre-conditions are met (or none is specified), or throws an error indicating
+ * that a pre-condition failed (the error's `statusCode` is set to `412`).
+ *
+ * NOTE: LocalStack ignores `IfMatch` and `IfNoneMatch` in the `params` object
+ * to S3 methods, so we must simulate 412 (PreconditionFailed) responses
+ * ourselves.  If LocalStack ever supports these pre-conditions, this function
+ * may be removed.
+ *
+ * @example s3().getObject(params).then(checkMatchPreconditions(params))
+ *
+ * @function
+ * @param {S3MatchParams} params - object optionally containing
+ * @returns {function} a function that compares the resolved result of an S3
+ *    request with the specified match parameters, and either returns the result
+ *    if pre-conditions are met (or unspecified), or throws an error (with
+ *    `statusCode` set to `412`) indicating a pre-condition failed
+ */
+const checkMatchPreconditions = (params: S3MatchParams) =>
```

Comment:
If this is here to just support testing via Localstack and we add a `getS3ObjectWithEtag` function, it should be easy to mock this behavior for the tests rather than needing it in this S3 client library itself?

Reply:
Good point. I hadn't considered that, but I'll see how that could be done.